src/GafferOSL/OSLObject.cpp

```diff
@@ -382,7 +391,11 @@ bool OSLObject::adjustBounds() const
 		return false;
 	}
 
-	ConstShadingEnginePtr s = shadingEngine( Context::current() );
+	// TODO - this is technically wrong, since in theory I think you could build a shader where all
+	// writes to P get constant folded out when the attribute substutitions aren't applied, but
+	// does actually write to P when the attributes are substituted.  Since adjustBounds is called
+	// in a global scope however, it's hard to do this properly.
```

Comment:
OK, if we really want to do this right:

ShadingEngine declares a special token STRING_PARAMETER_VALUE_UNKNOWN.  Evaluating shade() on a ShadingEngine containing these tokens throws an exception, but calls to hasDeformation yield correct results ( implemented using lockgeom=false ).

The getter for ShadingEngines inside OSLShader can then be modified to replace strings that require substitutions with this token when no attributes are provided ... I guess this might require a special sentinel value for the attributes that OSLObject::adjustBounds passes to this getter, to indicate "the attribute values are unknown" vs "the attribute values aren't provided" ( ie. useAttributes is set false ).

This approach seems like it's kinda complicated, but would provide correct behaviour.

Reply:
That does sound like it would work. But I'm willing to bet that this is enough of a corner case that it's not worth the effort or complexity at this point. My vote would be to just add a short comment linking to this discussion.